pkg/cmd/pulumi/cmd/project_sdk_link.go

```diff
@@ -0,0 +1,816 @@
+			err := os.MkdirAll(directory, 0o700)
+			if err != nil {
+				return err
+			}
+
+			jsonBytes, err := pkg.MarshalJSON()
+			if err != nil {
+				return err
+			}
+
+			pCtx, err := NewPluginContext(cwd)
+			if err != nil {
+				return fmt.Errorf("create plugin context: %w", err)
+			}
+			defer contract.IgnoreClose(pCtx.Host)
+			programInfo := plugin.NewProgramInfo(cwd, cwd, ".", nil)
+			languagePlugin, err := pCtx.Host.LanguageRuntime(language, programInfo)
+			if err != nil {
+				return err
+			}
+
+			loader := schema.NewPluginLoader(pCtx.Host)
+			loaderServer := schema.NewLoaderServer(loader)
+			grpcServer, err := plugin.NewServer(pCtx, schema.LoaderRegistration(loaderServer))
+			if err != nil {
+				return err
+			}
+			defer contract.IgnoreClose(grpcServer)
+
+			diags, err := languagePlugin.GeneratePackage(directory, string(jsonBytes), extraFiles, grpcServer.Addr(), nil, local)
+			if err != nil {
+				return err
+			}
+
+			// These diagnostics come directly from the converter and so _should_ be user friendly. So we're just
+			// going to print them.
+			cmdDiag.PrintDiagnostics(pCtx.Diag, diags)
+			if diags.HasErrors() {
+				// If we've got error diagnostics then package generation failed, we've printed the error above so
+				// just return a plain message here.
+				return errors.New("generation failed")
+			}
+
+			return nil
+		}
+	}
+
+	extraFiles := make(map[string][]byte)
+	if overlays != "" {
+		fsys := os.DirFS(filepath.Join(overlays, language))
+		err := fs.WalkDir(fsys, ".", func(path string, d fs.DirEntry, err error) error {
+			if err != nil || d.IsDir() {
+				return err
+			}
+
+			contents, err := fs.ReadFile(fsys, path)
+			if err != nil {
+				return fmt.Errorf("read overlay file %q: %w", path, err)
+			}
+
+			extraFiles[path] = contents
+			return nil
+		})
+		if err != nil {
+			return fmt.Errorf("read overlay directory %q: %w", overlays, err)
+		}
+	}
+
+	root := filepath.Join(out, language)
+	err = generatePackage(root, pkg, extraFiles)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Prints instructions for linking a locally generated SDK to an existing
+// project, in the absence of us attempting to perform this linking automatically.
+func DoLocalSdkLinking(
+	ws pkgWorkspace.Context, language string, root string, pkg *schema.Package, out string,
+) error {
+	switch language {
+	case "nodejs":
+		return linkNodeJsPackage(ws, root, pkg, out)
+	case "python":
+		return linkPythonPackage(ws, root, pkg, out)
+	case "go":
+		return linkGoPackage(root, pkg, out)
+	case "dotnet":
+		return linkDotnetPackage(root, pkg, out)
+	case "java":
+		return printJavaLinkInstructions(root, pkg, out)
+	default:
+		break
+	}
+	return nil
+}
+
+// Prints instructions for linking a locally generated SDK to an existing NodeJS
+// project, in the absence of us attempting to perform this linking automatically.
+func linkNodeJsPackage(ws pkgWorkspace.Context, root string, pkg *schema.Package, out string) error {
```

Comment:
This comment is probably outdated now. Same for any of the other `link*` helpers that are now automating stuff.